core/src/main/java/com/scalar/db/exception/transaction/TransactionException.java

```diff
@@ -1,13 +1,29 @@
 package com.scalar.db.exception.transaction;
 
+import java.util.Optional;
+import javax.annotation.Nullable;
+
 /** Base class for all exceptions thrown by the Transaction API. */
 public class TransactionException extends Exception {
 
-  public TransactionException(String message) {
+  @Nullable private final String transactionId;
```

Comment:
How about overriding `getMessage()` to include the transaction ID so that we can pass a message without transaction ID to constructer like this?

```java
  @Override
  public String getMessage() {
    return String.format("%s. transactionID:%s", super.getMessage(), transactionId);
  }
```

```java
    throw new TransactionException("The transaction already exists", transaction.getId());
```

Reply:
@komamitsu @Torch3333 

I have found a problem with this. We sometimes catch a transaction exception and wrap it with its `getMessage()` and rethrow it like this:
https://github.com/scalar-labs/scalardb/blob/3e5751acd96a5937464a70362d9d8d9610193928/core/src/main/java/com/scalar/db/transaction/consensuscommit/CommitHandler.java#L53-L56

In this case, if we override `TransactionException.getMessage()` that way, the exception message shows `transactionId` twice. And I haven't come up with a smart solution for this problem...

Do you have any idea?

Reply:
It is not a smart solution but I think its the simplest to check if the message does not already contains the transaction ID. 

```
  @Override
  public String getMessage() {
    String message = super.getMessage();
    String suffix = "transaction ID: " + transactionId;
    if (!message.endsWith(suffix)) {
      return String.format("%s. %s", message, suffix);
    } else {
      return message;
    }
  }
  ```
Another option would be to add another method, `getOriginalMessage()` to `TransactionException`,  to retrieve the original message without the transaction ID, but I prefer the solution above for its simplicity.

Reply:
Another option might be checking the existence of TransactionException-ish cause like this?

```java
    @Override
  public String getMessage() {
    if (hasTransactionExceptionInCauses()) {
      // The cause already contains transaction ID in its message
      return super.getMessage();
    }
    else {
      return String.format("%s. transactionID:%s", super.getMessage(), transactionId);
    }
  }
```

Reply:
Ah, I think I had the wrong idea. Let me see again...

Reply:
On second thought, it works as follows?

#### Case#1

```java
 if (e instanceof PreparationConflictException) { 
   throw new CommitConflictException(e.getMessage(), e, e.getTransactionId().orElse(null)); 
 } 
```

In this case, the new CommitConflictException instance should have
- message: "Preparation failed. transactionID=xxxx"
- cause: PreparationConflictException
  - message: "Preparation failed. transactionID=xxxx"
  - cause: Null or something
  - transactionId: xxxx
- transactionId: xxxx

With suggested the above `getMessage()`,
- CommitConflictException.getMessage() returns "Preparation failed. transactionID=xxxx"
  - as `hasTransactionExceptionInCauses()` returns true
- PreparationConflictException.getMessage() returns "Preparation failed. transactionID=xxxx"

#### Case#2

```java
 if (e instanceof IllegalArgumentException) { 
   throw new CommitConflictException(e.getMessage(), e, transactionId); 
 } 
```

In this case, the new CommitConflictException instance should have
- message: "An invalid parameter"
- cause: IllegalArgumentException
  - message: "An invalid parameter"
  - cause: Null or something
- transactionId: xxxx

With suggested the above `getMessage()`,
- CommitConflictException.getMessage() returns "An invalid parameter. transactionID=xxxx"
  - as `hasTransactionExceptionInCauses()` returns false
- IllegalArgumentException.getMessage() returns "An invalid parameter"

Reply:
BTW, IMO I don't think it's a so severe issue to happen to show `transactionId` twice

Reply:
It looks like we have one more case as follows:
https://github.com/scalar-labs/scalardb/blob/3e5751acd96a5937464a70362d9d8d9610193928/core/src/main/java/com/scalar/db/transaction/rpc/GrpcTransactionOnBidirectionalStream.java#L248

This is the case when we use ScalarDB Server. In this case, `error.getMessage()` should have the transaction ID, but the exception doesn't have any cause exceptions.

> BTW, IMO I don't think it's a so severe issue to happen to show transactionId twice

It looks like when we use ScalarDB Server (and ScalarDB Cluster), the transaction ID is always duplicated. I agree that it's not very severe, but I think we should handle it.

Example of the transaction ID duplicated messages:
```
The transaction already exists. transaction ID: 5d37796a-0e43-4ebb-8b4b-3da60d5a69a6. transaction ID: 5d37796a-0e43-4ebb-8b4b-3da60d5a69a6
```

So maybe, the Vincent's idea is more reasonable so far, though it's not perfect?

```java
  @Override
  public String getMessage() {
    String message = super.getMessage();
    String suffix = "transaction ID: " + transactionId;
    if (!message.endsWith(suffix)) {
      return String.format("%s. %s", message, suffix);
    } else {
      return message;
    }
  }
```

If there are no objections, I will go ahead with Vincent's idea.

Reply:
LGTM!

Reply:
@komamitsu @Torch3333 Thank you for the discussion! Fixed in 4e8dae7da1883defb20a0a4196be33a84c1bfbe9.

Please take a look again when you have time!